### ISSUE TITLE
Input types that contain itself

### ISSUE THREAD

**Issue by paussu** [microsoft/cppgraphqlgen#225]
Hi,

We are currently testing if we can upgrade our version from 3.x.x to 4.x.x and everything has been going smoothly except for one thing.

In our schema we have filter input types that refer to themselves.
```
input StringOperationFilterInput {
  _and: [StringOperationFilterInput!]
  _or: [StringOperationFilterInput!]
  _eq: String
  _neq: String
  _contains: String
  _ncontains: String
  _in: [String!]
  _nin: [String!]
  _startsWith: String
  _nstartsWith: String
  _endsWith: String
  _nendsWith: String
}

```

The new Schemagen doesn't like that and fails with:
`Input object cycle type: StringOperationFilterInput`

Any idea how this should be solved?

Thanks

**Comment by wravery** [comment]
Note to self: https://spec.graphql.org/October2021/#sec-Input-Objects.Circular-References

**Comment by wravery** [comment]
TL;DR; yes, it looks like the cycle detection is too aggressive. Nullable or list of self types should be allowed, only a non-nullable type should be blocked.

This has some unwelcome implications for the C++ code generation, though. A `std::vector<Self>` should be fine, but `std::optional<Self>` (for a nullable field) needs to be replaced with `std::unique_ptr<Self>` because `std::optional` stores the non-null value inline. A lot of code referencing them should port cleanly since they override the same operators, but there will likely be spots where that doesn't work without some cleanup. I suspect that v3.x without the cycle detection probably had a bug where it would generate invalid C++ using `std::optional<Self>` for a nullable, non-list self type field.

This'll also need a unit test or two, or at least a schema addition to validate the types that should work generate working C++ code.
